colossalai/auto_parallel/solver/op_handler/node_handler.py

```diff
@@ -28,13 +29,53 @@ def __init__(
         self.device_mesh = device_mesh
         self.strategies_vector = strategies_vector
 
-    def register_strategy(self) -> StrategiesVector:
+    def update_resharding_cost(self, strategy: ShardingStrategy_V2) -> None:
+        """
+        Compute the resharding costs and save the costs in the ShardingStrategy object.
+        """
+        # TODO: test this function when other handlers are ready
+        resharding_costs = {}
+        shape_consistency_manager = ShapeConsistencyManager()
+        for node in self.predecessor_node:
+            node_name = str(node)
+
+            # get the sharding specs for this node generated
+            # in its own node handler
+            assert hasattr(node, 'strategies_vector'), \
+                f'The predecessor node {node_name} has no strategy vector to compute the resharding cost.'
+            prev_strategy_vector = node.strategies_vector
+            prev_sharding_specs = [strategy.get_sharding_spec_by_name(node_name) for strategy in prev_strategy_vector]
+
+            # get the current sharding spec generated by this node handler
+            op_data = strategy.get_op_data_by_name(node_name)
+            current_sharding_spec = strategy.sharding_specs[op_data]
+
+            # create data structrure to store costs
+            if op_data not in resharding_costs:
+                resharding_costs[op_data] = {}
+
+            # for each sharding spec generated by the predecessor's node handler
+            # compute the resharding cost to switch to the sharding spec generated
+            # by the current node handler
+            for prev_sharding_spec in prev_sharding_specs:
+                fwd_cost = shape_consistency_manager.shape_consistency(prev_sharding_spec, current_sharding_spec)
+                bwd_cost = shape_consistency_manager.shape_consistency(current_sharding_spec, prev_sharding_spec)
```

Comment:
ShapeConsistencyManager will distinguish the mission type(training or inference) by a forward_only label, and returns a option related cost. Therefore, we don't have to compute fwd cost and bwd cost separately here.

Reply:
I see. I notice that ShapeConsistencyManager only returns a total cost. How to get the fwd and bwd cost then?

Reply:
![image](https://user-images.githubusercontent.com/72588413/192684440-c2e97386-dad3-47c4-a093-ee6c7b2ab242.png)

Reply:
I mean I want to store this cost as a TrainCycleItem so that solver can choose whether to consider the bwd cost.